feat(gateway): handle value types

steinroe opened the pull request:
Implements value type handling as discussed in #499

Leaves an open todo as described in find-value-types.js:
```
/**
 * TODO: it should be checked whether all types with the same name have the exact same fields,
 * incl. types, subtypes and nullability
 * If that is not the case, I would argue that throwing would be the best approach
 */
```

mcollina commented on test/gateway/value-types.js:
```diff
@@ -0,0 +1,410 @@
+  })
+})
+
+test('Should not register value types if field ', async (t) => {})
```
can you fill this in?

steinroe replied:
This seems to be a leftover. Or do you miss a test?

mcollina replied:
looks like it is, just drop it!

mcollina commented on test/gateway/value-types.js:
```diff
@@ -0,0 +1,410 @@
+    })
+  })
+
+  t.same(JSON.parse(usersRes.body), usersData)
```
```suggestion
  t.same(usersRes.json(), usersData)
```

and everywhere else.

mcollina commented on lib/gateway/find-value-types.js:
```diff
@@ -0,0 +1,54 @@
+      return a
+    }, {})
+  return Object.keys(uniq).filter((a) => uniq[a] > 1)
+}
```
Why can't we just put all names in a set instead?

steinroe replied:
Because `allTypes` is an array of `GraphQLObjectType` and not strings. Using the `GraphQLObjectType` instead of strings is required for `isObjectType` etc. to work properly.

mcollina commented on lib/gateway/find-value-types.js:
```diff
@@ -0,0 +1,54 @@
+ */
+function findValueTypes (allTypes) {
+  // Iterate all service defs and find all types by name that are used by more than one service
+  const duplicateNames = duplicates(allTypes.map(type => type.name))
```
Delete this map

steinroe commented:
Thanks for your quick feedback!

I changed the implementation for find-value-types into the following, but the test coverage for functions got reduced to 80% for that file and I cannot figure out why. Do you have an idea?

```js
const {
  isInterfaceType,
  isObjectType,
  isInputType
} = require('graphql')

function getDuplicateTypes (types) {
  const occurrences = {}
  const duplicates = []
  for (const type of types) {
    if (occurrences[type.name] === 1) {
      duplicates.push(type)
      occurrences[type.name] = 2
    } else {
      occurrences[type.name] = 1
    }
  }
  return duplicates
}

/**
 * Find all value types according to the definition given here:
 * https://www.apollographql.com/docs/federation/value-types/
 *
 * TODO: it should be checked whether all types with the same name have the exact same fields,
 * incl. types, subtypes and nullability
 * If that is not the case, I would argue that throwing would be the best approach
 */
function findValueTypes (allTypes) {
  // Is exposed by more than one service
  return getDuplicateTypes(allTypes)
    .filter(type => {
      // Has to be object, interface or input
      const isEligibleType = (isObjectType(type) || isInterfaceType(type) || isInputType(type))
      // Is not an entity
      const isNotEntity = !type.astNode.directives.find(directive => directive.name.value === 'key')
      return isEligibleType && isNotEntity
    })
    .map(type => type.name)
}

module.exports = {
  findValueTypes
}

```

mcollina commented on lib/gateway/find-value-types.js:
```diff
@@ -0,0 +1,54 @@
+const {
```
use strict is missing

mcollina commented:
I don't know, I'm sorry.

I would change `getDuplicateTypes` to:

```js
function getDuplicateTypes (types) {
  const unique = new Set()
  const duplicates = new Set()
  for (const type of types) {
    const name = type.name
    if (unique.has(name))
      duplicates.add(name)
    else
      unique.add(name)
  }
  return duplicates // no need to convert to array, we can use it with a `for...of` later
}
```

steinroe commented:
Thanks for your help. The coverage was missing because it never ran into `isNotEntity` in `return isEligibleType && isNotEntity`. I added a test for that and cleaned up.

mcollina reviewed (approved):
lgtm
